Plugins/JiraCommitHintPlugin/JiraCommitHintPlugin.cs

```diff
@@ -0,0 +1,157 @@
+                var localQuery = jdlQuery.CustomControl.Text;
+                var localStringFormat = stringTemplateSetting.CustomControl.Text;
+                var preview = GetMessageToCommit(localJira, localQuery, localStringFormat).FirstOrDefault();
+                MessageBox.Show($"First Task Preview:{Environment.NewLine}{Environment.NewLine}{(preview == null ? "[Empty Jira Query Result]" : preview.Text)}");
+            }
+            catch (Exception ex)
+            {
+                MessageBox.Show(ex.Message);
+            }
+        }
+
+        public override void Register(IGitUICommands gitUiCommands)
+        {
+            base.Register(gitUiCommands);
+            gitUiCommands.PostSettings += gitUiCommands_PostSettings;
+            gitUiCommands.PreCommit += gitUiCommands_PreCommit;
+            gitUiCommands.PostRepositoryChanged += gitUiCommands_PostRepositoryChanged;
+            UpdateJiraSettings();
+        }
+
+        private void UpdateJiraSettings()
+        {
+            jira = Jira.CreateRestClient(url[Settings], user[Settings], password[Settings]);
+            query = jdlQuery[Settings];
+            stringFormat = stringTemplateSetting[Settings];
+            if (btnPreview == null)
+                return;
+            btnPreview.Click -= btnPreviewClick;
+            btnPreview = null;
+        }
+
+        private void gitUiCommands_PostSettings(object sender, GitUIPostActionEventArgs e)
+        {
+            UpdateJiraSettings();
+        }
+
+        public override void Unregister(IGitUICommands gitUiCommands)
+        {
+            base.Unregister(gitUiCommands);
+            gitUiCommands.PreCommit -= gitUiCommands_PreCommit;
+            gitUiCommands.PostSettings -= gitUiCommands_PostSettings;
+            gitUiCommands.PostRepositoryChanged -= gitUiCommands_PostRepositoryChanged;
+            btnPreview.Click -= btnPreviewClick;
+        }
+
+        private void gitUiCommands_PreCommit(object sender, GitUIBaseEventArgs e)
+        {
+            currentMessages = GetMessageToCommit(jira, query, stringFormat);
+            foreach (var message in currentMessages)
+            {
+                e.GitUICommands.AddCommitTemplate(message.Title, () => message.Text);
+            }
+        }
+
+        private void gitUiCommands_PostRepositoryChanged(object sender, GitUIBaseEventArgs e)
+        {
+            if (currentMessages == null)
+                return;
+            foreach (var message in currentMessages)
+            {
+                e.GitUICommands.RemoveCommitTemplate(message.Title);
+            }
+            currentMessages = null;
+        }
+
+        private static JiraTaskDTO[] GetMessageToCommit(Jira jira, string query, string stringFormat)
+        {
+            try
+            {
+                return jira.Issues.GetIssuesFromJqlAsync(query).Result
```

Comment:
Of cause it can be made `async`. But the reason code look like this now is there was no async in old Atlassian SDK. As you can see, the plugin development was started about three years ago :) So, when I've updated Atlassian SDK, I've done just simple support of that SDK, as it was in older version. 
If I have some more free time, I will rewrite this too.